// neutrinos-module/esm2015/src/app/neutrinos-module/neutrinos.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule } from '@angular/forms';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { FlexLayoutModule } from '@angular/flex-layout';
import { NeutrinosSeedServicesModule } from 'neutrinos-seed-services';
import { MatIconModule } from '@angular/material/icon';
import { MatDialogModule } from '@angular/material/dialog';
import { MatButtonModule } from '@angular/material/button';
import { MatButtonToggleModule } from '@angular/material';
import { NCameraDirective } from './neutrinos-file/directives/n-camera.directive';
import { NDownloadDirective } from './neutrinos-file/directives/n-download.directive';
import { NScanDirective } from './neutrinos-file/directives/n-scanner.directive';
import { NVideoDirective } from './neutrinos-file/directives/n-video.directive';
import { NBarcodeDirective } from './neutrinos-file/directives/n-barcode.directive';
import { NTtsDirective } from './neutrinos-file/directives/n-tts.directive';
import { NFingerprintDirective } from './neutrinos-file/directives/n-fingerprint.directive';
import { NOcrDirective } from './neutrinos-file/directives/n-ocr.directive';
import { NShakeDirective } from './neutrinos-file/directives/n-shake.directive';
import { NFileUploadComponent } from './neutrinos-file/fileUpload/n-fileUpload.component';
import { NFileIOService } from './neutrinos-file/services/n-fileIO.service';
import { NAlertService } from './neutrinos-components/nAlertComponent/n-alert.service';
import { NAlertComponent } from './neutrinos-components/nAlertComponent/n-alert.component';
import { NToggleoptionsComponent } from './neutrinos-components/nToggleoptionsComponent/n-toggleoptions.component';
import { NSignatureComponent } from './neutrinos-components/nSignatureComponent/n-signature.component';
const EXPORTS = [
    NCameraDirective,
    NDownloadDirective,
    NScanDirective,
    NVideoDirective,
    NBarcodeDirective,
    NTtsDirective,
    NFingerprintDirective,
    NOcrDirective,
    NShakeDirective,
    NFileUploadComponent,
    NAlertComponent,
    NToggleoptionsComponent,
    NSignatureComponent,
];
let NeutrinosModule = class NeutrinosModule {
};
NeutrinosModule = tslib_1.__decorate([
    NgModule({
        imports: [
            HttpClientModule,
            FormsModule,
            CommonModule,
            MatIconModule,
            MatDialogModule,
            MatButtonModule,
            BrowserAnimationsModule,
            MatButtonToggleModule,
            NeutrinosSeedServicesModule,
            FlexLayoutModule,
        ],
        declarations: [
            EXPORTS,
        ],
        exports: [
            EXPORTS,
            NFileUploadComponent,
            NeutrinosSeedServicesModule,
        ],
        providers: [
            NFileIOService,
            NAlertService
        ],
        entryComponents: [
            NAlertComponent
        ]
    })
], NeutrinosModule);
export { NeutrinosModule };
