// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-barcode.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
// import { FileIOService } from '../services/filesystem.service';
import { NFileIOService } from '../services/n-fileIO.service';
import { NDirectiveOptions } from './directive.class';
var NBarcodeDirective = /** @class */ (function () {
    function NBarcodeDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NBarcodeDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getBarcode()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NBarcodeDirective.prototype.getBarcode = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getBarcode(_this.barcodeOptions).then(function (res) {
                return resolve(res);
            }).catch(function (err) { return reject(err); });
        });
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", NDirectiveOptions)
    ], NBarcodeDirective.prototype, "barcodeOptions", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], NBarcodeDirective.prototype, "onsuccess", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], NBarcodeDirective.prototype, "onerror", void 0);
    tslib_1.__decorate([
        HostListener('click'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NBarcodeDirective.prototype, "methodToHandleMouseClickAction", null);
    NBarcodeDirective = tslib_1.__decorate([
        Directive({
            selector: '[n-barcode]'
        }),
        tslib_1.__metadata("design:paramtypes", [NFileIOService])
    ], NBarcodeDirective);
    return NBarcodeDirective;
}());
export { NBarcodeDirective };
