// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-ocr.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
// import { FileIOService } from '../services/filesystem.service';
import { NFileIOService } from '../services/n-fileIO.service';
import { NDirectiveOptions } from './directive.class';
var NOcrDirective = /** @class */ (function () {
    function NOcrDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NOcrDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getOcr()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NOcrDirective.prototype.getOcr = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getOcr(_this.ocrOptions).then(function (res) {
                return resolve(res);
            }).catch(function (err) { return reject(err); });
        });
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", NDirectiveOptions)
    ], NOcrDirective.prototype, "ocrOptions", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], NOcrDirective.prototype, "onsuccess", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], NOcrDirective.prototype, "onerror", void 0);
    tslib_1.__decorate([
        HostListener('click'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NOcrDirective.prototype, "methodToHandleMouseClickAction", null);
    NOcrDirective = tslib_1.__decorate([
        Directive({
            selector: '[n-ocr]'
        }),
        tslib_1.__metadata("design:paramtypes", [NFileIOService])
    ], NOcrDirective);
    return NOcrDirective;
}());
export { NOcrDirective };
