// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-file/directives/n-shake.directive.js
import * as tslib_1 from "tslib";
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { HostListener } from '@angular/core';
// import { FileIOService } from '../services/filesystem.service';
import { NFileIOService } from '../services/n-fileIO.service';
import { NDirectiveOptions } from './directive.class';
var NShakeDirective = /** @class */ (function () {
    function NShakeDirective(fsv) {
        this.fsv = fsv;
        this.onsuccess = new EventEmitter();
        this.onerror = new EventEmitter();
    }
    NShakeDirective.prototype.methodToHandleMouseClickAction = function () {
        var _this = this;
        this.getShake()
            .then(function (sucess) { return _this.onsuccess.emit(sucess); })
            .catch(function (error) { return _this.onerror.emit(error); });
    };
    NShakeDirective.prototype.getShake = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            _this.fsv.getShake(_this.shakeOptions).then(function (res) {
                return resolve(res);
            }).catch(function (err) { return reject(err); });
        });
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", NDirectiveOptions)
    ], NShakeDirective.prototype, "shakeOptions", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], NShakeDirective.prototype, "onsuccess", void 0);
    tslib_1.__decorate([
        Output(),
        tslib_1.__metadata("design:type", EventEmitter)
    ], NShakeDirective.prototype, "onerror", void 0);
    tslib_1.__decorate([
        HostListener('click'),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], NShakeDirective.prototype, "methodToHandleMouseClickAction", null);
    NShakeDirective = tslib_1.__decorate([
        Directive({
            selector: '[n-shake]'
        }),
        tslib_1.__metadata("design:paramtypes", [NFileIOService])
    ], NShakeDirective);
    return NShakeDirective;
}());
export { NShakeDirective };
