// neutrinos-module/esm5/src/app/neutrinos-module/neutrinos-components/nAlertComponent/n-alert.component.js
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { MatDialogRef } from '@angular/material';
var NAlertComponent = /** @class */ (function () {
    function NAlertComponent(dialogRef) {
        this.dialogRef = dialogRef;
        this.messageContent = NAlertComponent_1.message;
        this.titleContent = NAlertComponent_1.title;
    }
    NAlertComponent_1 = NAlertComponent;
    var NAlertComponent_1;
    NAlertComponent.title = '';
    NAlertComponent.message = '';
    NAlertComponent = NAlertComponent_1 = tslib_1.__decorate([
        Component({
            selector: 'n-alert-dialog',
            template: "\n<div class=\"alertContainer\" fxLayout=\"column\">\n  <h4 mat-dialog-title>{{titleContent}}</h4>\n  <span class=\"msgContainer\" >{{messageContent}}</span>\n  <button mat-raised-button (click)=\"dialogRef.close()\">Close</button>\n</div>\n",
            styles: [".alertContainer{height:auto;width:auto}.msgContainer{margin-bottom:.5em}"]
        }),
        tslib_1.__metadata("design:paramtypes", [MatDialogRef])
    ], NAlertComponent);
    return NAlertComponent;
}());
export { NAlertComponent };
